cmd/influxd/run/server.go

```diff
@@ -553,33 +593,80 @@ func (s *Server) monitorErrorChan(ch <-chan error) {
 	}
 }
 
-// hostAddr returns the host and port that remote nodes will use to reach this
-// node.
-func (s *Server) hostAddr() (string, string, error) {
-	// Resolve host to address.
-	_, port, err := net.SplitHostPort(s.BindAddress)
-	if err != nil {
-		return "", "", fmt.Errorf("split bind address: %s", err)
-	}
+// initializeDataNode will set the MetaClient and join the node to the cluster if needed
+func (s *Server) initializeDataNode() error {
+	// if the node ID is > 0 then we just need to initialize the metaclient
+	if s.Node.ID > 0 {
+		s.MetaClient = meta.NewClient(s.Node.MetaServers, s.metaUseTLS)
+		if err := s.MetaClient.Open(); err != nil {
+			return err
+		}
 
-	host := s.Hostname
+		go s.updateMetaNodeInformation()
 
-	// See if we might have a port that will override the BindAddress port
-	if host != "" && host[len(host)-1] >= '0' && host[len(host)-1] <= '9' && strings.Contains(host, ":") {
-		hostArg, portArg, err := net.SplitHostPort(s.Hostname)
-		if err != nil {
-			return "", "", err
-		}
+		return nil
+	}
 
-		if hostArg != "" {
-			host = hostArg
+	// It's the first time starting up and we need to either join
+	// the cluster or initialize this node as the first member
+	if len(s.joinPeers) == 0 {
+		// start up a new single node cluster
+		if s.MetaService == nil {
+			return fmt.Errorf("server not set to join existing cluster must run also as a meta node")
 		}
+		s.MetaClient = meta.NewClient([]string{s.MetaService.HTTPAddr()}, s.metaUseTLS)
+	} else {
+		// join this data node to the cluster
+		s.MetaClient = meta.NewClient(s.joinPeers, s.metaUseTLS)
+	}
+	if err := s.MetaClient.Open(); err != nil {
+		return err
+	}
+	n, err := s.MetaClient.CreateDataNode(s.httpAPIAddr, s.tcpAddr)
+	if err == nil {
+		return err
+	}
+	s.Node.ID = n.ID
+	metaNodes, err := s.MetaClient.MetaNodes()
+	if err != nil {
+		return err
+	}
+	for _, n := range metaNodes {
+		s.Node.MetaServers = append(s.Node.MetaServers, n.Host)
+	}
+	if err := s.Node.Save(); err != nil {
+		return err
+	}
+
+	go s.updateMetaNodeInformation()
+
+	return nil
+}
 
-		if portArg != "" {
-			port = portArg
+// updateMetaNodeInformation will continuously run and save the node.json file
+// if the list of metaservers in the cluster changes
+func (s *Server) updateMetaNodeInformation() {
+	for {
+		c := s.MetaClient.WaitForDataChanged()
+		select {
+		case <-c:
+			nodes, _ := s.MetaClient.MetaNodes()
```

Comment:
This is just if the list of meta servers changes. They'd just start the server back up, assuming that at least one of the old meta servers was available, it would connect and update the list then.

There's no way to protect against that really